packages/coinjoin/src/client/CoinjoinPrison.ts

```diff
@@ -1,48 +1,99 @@
+import { TypedEmitter } from '@trezor/utils/lib/typedEventEmitter';
+
+import { CoinjoinPrisonInmate, CoinjoinPrisonEvents } from '../types/client';
 import { WabiSabiProtocolErrorCode } from '../enums';
 
-export interface PrisonInmate {
-    id: string; // AccountUtxo/Alice.outpoint or AccountAddress scriptPubKey
-    sentenceStart: number;
-    sentenceEnd: number;
-    errorCode?: WabiSabiProtocolErrorCode | 'blameOf';
-    reason?: string;
-    roundId?: string;
-}
+export type DetainObject =
+    | {
+          outpoint: string;
+          accountKey: string;
+      }
+    | {
+          address: string;
+          accountKey: string;
+      }
+    | {
+          accountKey: string;
+      };
 
-export interface PrisonOptions {
+export interface DetainOptions {
     roundId?: string;
-    errorCode?: PrisonInmate['errorCode'];
-    reason?: PrisonInmate['reason'];
+    errorCode?: CoinjoinPrisonInmate['errorCode'];
+    reason?: CoinjoinPrisonInmate['reason'];
     sentenceEnd?: number;
 }
 
 // Errored or currently registered inputs and addresses are sent here
 // inspiration: WalletWasabi/WabiSabi/Backend/Banning/Prison.cs
 
-export class CoinjoinPrison {
-    inmates: PrisonInmate[] = [];
+export class CoinjoinPrison extends TypedEmitter<CoinjoinPrisonEvents> {
+    inmates: CoinjoinPrisonInmate[] = [];
+    private changeEventTimeout: ReturnType<typeof setTimeout> | undefined;
+
+    constructor(initialState: CoinjoinPrisonInmate[] = []) {
+        super();
+        this.inmates = initialState;
+    }
+
+    private dispatchChange() {
+        // throttle change events. might be emitted one after another (example: multiple detentions after round end)
+        clearTimeout(this.changeEventTimeout);
+        this.changeEventTimeout = setTimeout(
```

Comment:
not cleared on release?

Reply:
[db78df9](https://github.com/trezor/trezor-suite/pull/8549/commits/db78df9ef35854458380ab9c2ce1154d8c93dff4)